perf test a stress test to profile CPU load of podman

kunalkushwaha opened the pull request:
perftest uses libpod as golang library and perform various operations on containers.
This can be used as handy tool to stress test and profile CPU usage of libpod. 

Motivation for this tool is have tool, which can generate profiling data, which can be consumed by standard tools like `go tool pprof`. This could he helpful during integration of CRI-O and libpod.

Currently only stress test is `SingleThreadedStressTest`, but can be extended for parallel tests like `MultiThreadedStressTest` and extend operation for pods also.

Signed-off-by: Kunal Kushwaha <kushwaha_kunal_v7@lab.ntt.co.jp>

vrothberg commented on contrib/perftest/main.go:
```diff
@@ -0,0 +1,261 @@
+func (data *profileData) printProfiledData(testCount float64) {
+
+	fmt.Printf("\nProfile data\n\n")
+	fmt.Printf("\tCreate\tStart\tStop\tDelete\n")
```
I recommend using the [tabwriter](https://golang.org/pkg/text/tabwriter/) package instead. It's working here as the digits are limited, but it's harder to extend.

vrothberg commented on contrib/perftest/README.md:
````diff
@@ -0,0 +1,51 @@
+Build:
+
+```
+# cd $GOPATH/src/github.com/containers/libpod/contrib/perftest
````
An integration into the base Makefile would be nice, for example:
- `make perftest` for building
- `make run-perfest` for running

vrothberg commented on contrib/perftest/main.go:
```diff
@@ -0,0 +1,261 @@
+		return
+	}
+
+	logrus.SetLevel(logrus.ErrorLevel)
```
A `--log-level` flag might come in handy for debugging the tool.

kunalkushwaha replied:
done :)

TomSweeneyRedHat commented on contrib/perftest/main.go:
```diff
@@ -0,0 +1,261 @@
+	imageName := ""
+
+	testImageName := flag.String("image", "docker.io/library/alpine:latest", "image-name to be used for test")
+	testRunCount := flag.Int("count", 50, "count of loop counter for test")
```
I see the help message for the count, but it looks like you've 50 hard coded here?  Is that coming with future changes?

kunalkushwaha replied:
50 is default value, will be overridden if value is provided for flag.
e.g ``# perftest --count 10``

rhatdan replied:
Maybe make it a constant like defaultCount, so that we don't have magic numbers floating around.

TomSweeneyRedHat replied:
@kunalkushwaha right, but I was expecting to see a variable like inputCount here in line 50 rather than 50.

rhatdan commented:
bot, retest this please

mheon commented:
@kunalkushwaha Please rebase to pick up CI fixes

rhatdan commented:
/approve

kunalkushwaha commented:
This requires /ok-to-test

vrothberg commented:
/ok-to-test

@kunalkushwaha, could you answer the open question from 
https://github.com/containers/libpod/pull/1757#pullrequestreview-171450880?

Those were asking whether the proposed tool or a shell script is a better solution to do benchmarking and profiling. You've invested a great deal of time in this PR since the and it would be bad if the common conclusion is to drop this PR and to go for a shell script instead.

kunalkushwaha commented:
@vrothberg : I had updated the PR message as per your suggestion. Sorry didn't updated, you that I had updated.

----
Motivation for this tool is have a tool, which can generate profiling data, which can be consumed by standard tools like go tool pprof. This could he helpful during integration of CRI-O and libpod. These things are not possible with shell scripts.

Currently only stress test is `SingleThreadedStressTest`, but can be extended for parallel tests like `MultiThreadedStressTest` and extend operation for pods also. 

I kept it small for initial PR, so its  easy to review and merge.

kunalkushwaha commented:
@rhatdan This PR is complete.
 Additional test as can be added later per requirements.



Some of Tests which have failed are due to ``VarlinkClient`` tests. I same errors are occurring for all pull requests.

```
======================================================================
ERROR: test_999_remove (test.test_pods_ctnrs.TestPodsCtnrs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/go/src/github.com/containers/libpod/contrib/python/podman/podman/libs/pods.py", line 78, in remove
    results = podman.RemovePod(self._ident, force)
  File "/usr/lib/python3.6/site-packages/varlink/client.py", line 84, in _wrapped
    return self._call(method.name, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/varlink/client.py", line 141, in _call
    (message, more) = self._next_varlink_message()
  File "/usr/lib/python3.6/site-packages/varlink/client.py", line 111, in _next_varlink_message
    raise e
varlink.error.VarlinkError: {'parameters': {'reason': 'cgroups: cgroup mountpoint does not exist'}, 'error': 'io.podman.ErrorOccurred'}
```

rhatdan commented:
@kunalkushwaha Could you rebase and repush, Seems lots of fixes have been moved into CI, which might require a rebase.

kunalkushwaha commented:
All issues related to the PR are fixed.
-  Updated `vendor.conf` dependencies in `/vendor` i.e. [`github.com/pkg/profile`](github.com/pkg/profile)
-  Since, there is change in signature of function `libpod.WithNetNS`, updated the test for same.

Currently `full_vm_testing` fails due to some integration test related to `rootless tests`, which are not related to this PR.

mheon commented:
Cirrus is exploding - seems like only the rootless tests. I think I remember @cevich talking about failing rootless tests in IRC today - was that on master?

kunalkushwaha commented:
Yes, tests are failing on HEAD branch.

cevich commented:
/me looking...

cevich commented:
...Hmmm other tests failing as well in this PR's case.  I had thought the failures I was seeing were limited to my newer VM images + old runc + needing to passthrough ``$XDG_RUNTIME_DIR``.  Looking at the long list of job failures in Cirrus to see if this is more systemic...

cevich commented:
...okay, interesting.  The rootless tests do seem to be failing, including on [master as indicated by the status badge](https://github.com/containers/libpod#continuous-integration-).  I think this will be fixed by #1788, and it seems to be really close to passing tests.

mheon commented:
This LGTM... Let's see about getting this merged. Might need another rebase to fix the tests?

cevich commented:
looking

cevich commented:
ya, need rebase for PASS.  It's failing on the 'rootless' tests we fixed (last?) week.

kunalkushwaha commented:
Need to fix this PR with changes regarding https://github.com/containers/libpod/pull/1904/files#diff-76939214d3033492b329833fb18b7088R574. Will look into tomorrow.

kunalkushwaha commented:
I need help regarding this issue.

While creating container, I am getting this error 
`FATA[0000] no volume with name roc found in database: no such volume` at https://github.com/containers/libpod/blob/master/libpod/boltdb_state_internal.go#L574

For containerSpec settings are donw in this PR at https://github.com/containers/libpod/pull/1757/files#diff-0ca8a76a233793f50c7f50145a2de951R124

and creation is done with [ImageRootFs](https://github.com/containers/libpod/pull/1757/files#diff-0ca8a76a233793f50c7f50145a2de951R192)

The mount points are set to default one as below.

```
[
  { /proc proc proc      [ nosuid noexec nodev  ]  }   
  { /dev tmpfs tmpfs      [ nosuid strictatime mode=755 size=65536k  ]  }  
  {  /dev/pts devpts devpts      [ nosuid noexec newinstance ptmxmode=0666 mode=0620 gid=5  ]  }   
  {  /dev/shm tmpfs shm      [  nosuid noexec nodev mode=1777 size=65536k   ]  } 
  { /dev/mqueue mqueue mqueue      [  nosuid noexec nodev   ]  }  
  { /sys sysfs sysfs      [ nosuid noexec nodev ro  ]  }   
  { /sys/fs/cgroup cgroup cgroup      [  rprivate nosuid noexec nodev relatime ro  ] }
]
```
This error started coming after #1906 . Is anything changed with this PR, which is required to set with container while creation?

cc @rhatdan @mheon @umohnani8

cevich commented:
I don't have a direct answer to your question, but wonder if the unittest failure indicates anything related:

```
# github.com/containers/libpod/contrib/perftest
contrib/perftest/main.go:102:4: cannot use !true (type bool) as type "github.com/containers/libpod/vendor/github.com/containers/image/types".OptionalBool in field value
contrib/perftest/main.go:106:45: too many arguments in call to client.ImageRuntime().New
	have (context.Context, string, string, string, *os.File, *image.DockerRegistryOptions, image.SigningOptions, bool, bool)
	want (context.Context, string, string, string, io.Writer, *image.DockerRegistryOptions, image.SigningOptions, bool)
```

I see that in both the Fedora and Ubuntu results.

mheon commented:
@kunalkushwaha It looks like that's a side effect of the image volumes work from @umohnani8 - it seems like there's a volume somewhere in your create config that has a relative path, which I don't believe is supported since the image volumes work landed.

I don't see any volumes named "roc", though... @umohnani8 Could this be an issue with image volumes interacting with named volumes? Only place I can think an unexpected volume would pop up from

kunalkushwaha commented:
@mheon I feel this __roc__ volume is /proc , which is somehow getting intial letter is getting truncated.

mheon commented:
That would make sense...

Alright, I think I know what needs to happen here, but it could end up being a lot of code. We need to actually record the named volumes in the container config, as opposed to just adding them to the spec, which will allow us to more accurately detect whether we have named volumes in the container.

kunalkushwaha commented:
@mheon I got the issue. and currently testing the fix with all test cases. Will make PR. 
thanks for pointing to namedVolume.

mheon commented:
@kunalkushwaha Ack - glad to hear!

kunalkushwaha commented:
_ctr.runtime.config.VolumePath_ is empty, due to which, condition satisfy and _proc_ is manipulated.
check with len(ctr.runtime.config.VolumePath) > 0 && .. should fix it.

https://github.com/containers/libpod/blob/master/libpod/boltdb_state_internal.go#L569

kunalkushwaha commented:
The test failed in CI are not related to this PR. the failing for all outstanding PRs.

```
test/test_runner.sh: line 51: 11358 Killed                  $PODMAN varlink --timeout=0 ${PODMAN_HOST} >> /tmp/test_podman.output 2>&1
/tmp/busybox.log =====
cbfc0b3c167a44d1b05e9fa4d0e9ea01f43c6a65d0530eaedf2980397c1857ca

umount /tmp/podman.IXuMHBWADR/crio/vfs-containers/c99785ca713cff485e9ef8ba40102ff46a986eaef0259266c39b962260626c9c/userdata/shm /run/netns/cni-ea0d12b5-774c-f776-f870-ea7cfbeaba0e /run/netns/cni-ea0d12b5-774c-f776-f870-ea7cfbeaba0e /run/netns/cni-16aefb7a-9439-7804-605f-9ed301fb8402 /run/netns/cni-16aefb7a-9439-7804-605f-9ed301fb8402 /tmp/podman.IXuMHBWADR/crio/vfs-containers/dcd72a43881f1140e99b6d5bc7dfcd2cb03e02a29bede942bb83a58c447b9d4c/userdata/shm 
umount: /run/netns/cni-ea0d12b5-774c-f776-f870-ea7cfbeaba0e: not mounted.
umount: /run/netns/cni-16aefb7a-9439-7804-605f-9ed301fb8402: not mounted.
umount /run/netns 
Makefile:16: recipe for target 'integration' failed
make[1]: *** [integration] Error 1
make[1]: Leaving directory '/var/tmp/go/src/github.com/containers/libpod/contrib/python/podman'
Makefile:189: recipe for target 'clientintegration' failed
make: *** [clientintegration] Error 2

Exit status: 2
```

cevich commented:
> 
The test failed in CI are not related to this PR. the failing for all outstanding PRs.

Ugg, I think that's the mount/cgroups error from the python tests, yes?  That's a known flake, though frustrating, just smash the 're-test' button if you see it.

It's a "This Should Never Happen" kind of things.  We spent days trying to reproduce / figure that problem out. Somehow/why it happens in GCE more than on our laptops  :confounded:

rhatdan commented:
bot, retest this please

kunalkushwaha commented:
@rhatdan The test failing is not related to this PR.

A request, as I will be on leave for 2 weeks from next week, So if anything requires for this PR, I can fix this week.

cevich commented:
@kunalkushwaha rebase from master and that failing test will go away.  Hopefully replaced by a new, non-failing test (but I can help if it dies).

kunalkushwaha commented:
@cevich fedora environment setup have issues 

``` bash
export LD_LIBRARY_PATH="/usr/local/lib${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}"
>>>>> OVERRIDING RUNC WITH https://kojipkgs.fedoraproject.org/packages/runc/1.0.0/55.dev.git578fe65.fc28/x86_64/runc-1.0.0-55.dev.git578fe65.fc28.x86_64.rpm <<<<<
Copr repo for gce-oslogin owned by ngompa        19 kB/s | 4.9 kB     00:00    
Fedora 28 - x86_64 - Updates                     22 MB/s |  28 MB     00:01    
Error: Failed to synchronize cache for repo 'fedora'
```

cevich commented:
@kunalkushwaha that's probably just some networking hiccup, they happen sometimes.  Just re-run if you see that.  I'll hit it for you this time, next time it'll cost you $0.50 :smile:

kunalkushwaha commented:
bot, retest

mheon commented:
CI is green
LGTM
@rhatdan @baude @TomSweeneyRedHat PTAL and merge

rhatdan commented:
LGTM
@kunalkushwaha Nice work, and good perseverance on getting this through.
